src/orderbook/index.ts

```diff
@@ -0,0 +1,258 @@
+    const cached = this.cache.get(blockNumber)
+    if (cached) {
+      return cached
+    } else {
+      const block = await this.web3.eth.getBlock(blockNumber)
+      const batchId = ~~(Number(block.timestamp) / BATCH_DURATION)
+      this.cache.set(blockNumber, batchId)
+      return batchId
+    }
+  }
+}
+
+/**
+ * The streamed orderbook that manages incoming events, and applies them to the
+ * account state.
+ */
+export class StreamedOrderbook {
+  private readonly batches: BatchIds;
+  private readonly state: AccountState;
+
+  private constructor(
+    private readonly web3: Web3,
+    private readonly contract: BatchExchange,
+    private readonly startBlock: number,
+    private readonly options: OrderbookOptions,
+  ) {
+    this.batches = new BatchIds(web3)
+    this.state = new AccountState(this.batches, options)
+  }
+
+  /**
+   * Create and return a new streamed orderbook.
+   *
+   * @remarks
+   * This method returns a promise that resolves once all past events have been
+   * applied to the current account state and the orderbook.
+   *
+   * @param web3 - The web3 provider to use.
+   * @param options - Optional settings for tweaking the streamed orderbook.
+   */
+  static async init(
+    web3: Web3,
+    options: Partial<OrderbookOptions> = {},
+  ): Promise<StreamedOrderbook> {
+    const [contract, tx] = await batchExchangeDeployment(web3)
+    const orderbook = new StreamedOrderbook(
+      web3,
+      contract,
+      tx.blockNumber,
+      { ...DEFAULT_ORDERBOOK_OPTIONS, ...options },
+    )
+
+    await orderbook.applyPastEvents()
+    return orderbook
+  }
+
+  /**
+   * Apply all past events to the account state by querying the node for past
+   * events with multiple queries to retrieve each block page at a time.
+   */
+  private async applyPastEvents(): Promise<void> {
+    const endBlock = this.options.endBlock ||
+      await this.web3.eth.getBlockNumber()
+
+    let pendingEvents: EventData[] = []
+    for (
+      let fromBlock = this.startBlock;
+      fromBlock < endBlock;
+      fromBlock += this.options.blockPageSize
+    ) {
+      // NOTE: `getPastEvents` block range is inclusive.
+      const toBlock = Math.min(
+        fromBlock + this.options.blockPageSize - 1,
+        endBlock,
+      )
+      console.error(`fetching page ${fromBlock}-${toBlock}`)
+
+      const events = await this.contract.getPastEvents(
+        'allEvents',
+        { fromBlock, toBlock },
+      )
+
+      const allEvents = pendingEvents.concat(events)
+      pendingEvents = await this.applyEventsUntilLastBatch(allEvents)
+    }
+  }
+
+  /**
+   * Finds the last complete batch in the specified events and applies all
+   * events up until that point to the account state.
+   *
+   * @param allEvents - All events to potentially apply.
+   * @returns A promise that resolves to the remaining events that could not be
+   * applied to the account state as there are potentially missing events from
+   * that batch.
```

Comment:
Mention that the order of those events stays the same. That makes it clear why we reverse at the end of the function.